pallets/loans/src/lib.rs

```diff
@@ -770,6 +766,174 @@ pub mod pallet {
+				TransactionOutcome::Rollback(result)
+			})?;
+
+			T::ChangeGuard::note(
+				pool_id,
+				Change::TransferDebt(from_loan_id, to_loan_id, repaid_amount, borrow_amount).into(),
+			)?;
+
+			Ok(())
+		}
+
+		/// Transfer debt from one loan to another loan,
+		/// repaying from the first loan and borrowing the same amount from the
+		/// second loan. `from_loan_id` is the loan used to repay.
+		/// `to_loan_id` is the loan used to borrow.
+		/// The repaid and borrow amount must match.
+		#[pallet::weight(T::WeightInfo::apply_transfer_debt(T::MaxActiveLoansPerPool::get()))]
+		#[pallet::call_index(12)]
+		pub fn apply_transfer_debt(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			change_id: T::Hash,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let Change::TransferDebt(from_loan_id, to_loan_id, repaid_amount, borrow_amount) =
+                Self::get_released_change(pool_id, change_id)? else {
+                    Err(Error::<T>::UnrelatedChangeId)?
+                };
+
+			let (amount, _count) = Self::transfer_debt_action(
+				&who,
+				pool_id,
+				from_loan_id,
+				to_loan_id,
+				repaid_amount,
+				borrow_amount,
+				true,
+			)?;
+
+			Self::deposit_event(Event::<T>::DebtTransferred {
+				pool_id,
+				from_loan_id,
+				to_loan_id,
+				amount,
+			});
+
+			Ok(())
+		}
+	}
+
+	// Loan actions
+	impl<T: Config> Pallet<T> {
+		fn borrow_action(
+			who: &T::AccountId,
+			pool_id: T::PoolId,
+			loan_id: T::LoanId,
+			amount: &PrincipalInput<T>,
+			permissionless: bool,
+		) -> Result<u32, DispatchError> {
+			Ok(match CreatedLoan::<T>::take(pool_id, loan_id) {
+				Some(created_loan) => {
+					if !permissionless {
+						Self::ensure_loan_borrower(who, created_loan.borrower())?;
+					}
+
+					let mut active_loan = created_loan.activate(pool_id, amount.clone())?;
+					active_loan.borrow(amount, pool_id)?;
+
+					Self::insert_active_loan(pool_id, loan_id, active_loan)?
+				}
+				None => {
+					Self::update_active_loan(pool_id, loan_id, |loan| {
+						if !permissionless {
+							Self::ensure_loan_borrower(who, loan.borrower())?;
+						}
+
+						loan.borrow(amount, pool_id)
+					})?
+					.1
+				}
+			})
+		}
+
+		fn repay_action(
+			who: &T::AccountId,
+			pool_id: T::PoolId,
+			loan_id: T::LoanId,
+			amount: &RepaidInput<T>,
+			permissionless: bool,
+		) -> Result<(RepaidInput<T>, u32), DispatchError> {
+			Self::update_active_loan(pool_id, loan_id, |loan| {
+				if !permissionless {
+					Self::ensure_loan_borrower(who, loan.borrower())?;
+				}
+
+				loan.repay(amount.clone(), pool_id)
+			})
+		}
+
+		fn transfer_debt_action(
+			who: &T::AccountId,
+			pool_id: T::PoolId,
+			from_loan_id: T::LoanId,
+			to_loan_id: T::LoanId,
+			repaid_amount: RepaidInput<T>,
+			borrow_amount: PrincipalInput<T>,
+			permissionless: bool,
+		) -> Result<(T::Balance, u32), DispatchError> {
+			ensure!(
+				from_loan_id != to_loan_id,
+				Error::<T>::TransferDebtToSameLoan
+			);
```

Comment:
Just to double-check @hieronx, could it make sense to allow transferring debt to the same loan? i.e. for paying the interest with the amount borrowed from the principal.

Maybe I'm being too strict with this restriction